sound/soc/sof/intel/hda-dsp.c

```diff
@@ -381,6 +406,100 @@ int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
 	return ret;
 }
 
+static int hda_dsp_set_D3_state(struct snd_sof_dev *sdev,
+				const struct sof_dsp_power_state *target_state)
+{
+	/*
+	 * Sanity check for illegal state transitions
+	 * The only allowed transition is: D0I0 -> D3
+	 */
+	switch (sdev->dsp_power_state.state) {
+	case SOF_DSP_D0:
+		if (sdev->dsp_power_state.substate != SOF_HDA_DSP_D0I0) {
+			dev_err(sdev->dev,
+				"error: transition from D0I3 -> D3 not allowed\n");
+			return -EINVAL;
+		}
+		break;
+	case SOF_DSP_D3:
+		/* already in D3 */
+		return 0;
```

Comment:
can this actually happen that you ask for a transition to D3 while already being in D3? Looks like something very wrong in the state machine, no?

Reply:
No this cannot happen. Let me think of a better way to show the state machine in this case.